### PR TITLE
feat(ngMocks): add expectRoute and whenRoute shortcuts with colon param matching

### PR DESCRIPTION
**Pull request by dmaloneycalu**
Adds a `whenRoute()` and `expectRoute()` function to `$httpBackend` service to make mocking out responses similar to writing API routes on a server. Also includes colon delimited param matching so that expect and when handlers can match requests in the form of `/user/:id`, identical to the way ngRoute matches urls in an Angular application.

Here's an [example file](https://gist.github.com/dmaloneycalu/cb258e2f3b987fd687b2) that reflects one of the use cases for the feature (paged API responses).

### CODE REVIEW

**Comment by dmaloneycalu** [comment]
CLA signed and builds passing. Any feedback here?

**Comment by dmaloneycalu** [comment]
Any thoughts or suggestions on this? Potentially a 1.4.x or 1.5 addition?

@vojtajina @petebacondarwin @matsko

**Comment by dmaloneycalu** [comment]
Have not seen much action on the Angular repo lately, wanted to give it a bump

**Comment by dmaloneycalu** [comment]
I can rebase this on master and squash the commits if necessary. Still haven't received any feedback. @angular

**Comment by petebacondarwin** [comment]
I like the look of this @dmaloneycalu please do rebase. We probably won't be able to get to it until after AngularConnect now.

**Comment by gkalpak** [comment]
The tests might also make good use of the [`they()`][1] helper.

[1]: https://github.com/angular/angular.js/blob/291d7c467fba51a9cb89cbeee62202d51fe64b09/test/helpers/privateMocks.js#L22

**Comment by gkalpak** [comment]
BTW, the real benefit I see in this change comes from the added `keys` parameter. (Matching routes could be achieved with RegExps before.)

So, we need docs and tests for that :)

**Comment by dmaloneycalu** [comment]
@petebacondarwin can do.

@gkalpak I'll add additional tests for the parameter addition and rewrite some of the tests to take advantage of the `they()` helper

**Comment by dmaloneycalu** [comment]
@petebacondarwin @gkalpak 

- squashed commits
- added additional tests for params matching outside of `whenRoute` and `expectRoute`
- fixed a few documentation lines I was missing

File: src/ngMock/angular-mocks.js [hunk 1]
```diff
@@ -1251,20 +1251,21 @@ function createHttpBackendMock($rootScope, $timeout, $delegate, $browser) {
    *   data string and returns true if the data is as expected.
    * @param {(Object|function(Object))=} headers HTTP headers or function that receives http header
    *   object and returns true if the headers match the current definition.
+   * @param {(Array)=} keys Array of keys to assign to regex matches in the respond params for the url.
```
Comment (petebacondarwin) [inline src/ngMock/angular-mocks.js]: I think that this needs a bit more explanation for people coming to this documentation without looking at the code.  Perhaps we could have a single section in the ngMock.$httpBackend description that explains what can now be done with regexes and keys then simply refer each fo these parameter descriptions to that section?

Comment (dmaloneycalu) [inline src/ngMock/angular-mocks.js]: I'll take a stab at this and notify you when I've pushed it.

**Comment by dmaloneycalu** [comment]
@petebacondarwin
- Added a section to the documentation
- Reordered code so documentation for `whenRoute` and `expectRoute` were at bottom of shortcuts list
- Pointed method param declarations to see above for further explanation

I'm not an English professor, so feel free to let me know if anything isn't well defined enough

**Comment by petebacondarwin** [comment]
Thanks @dmaloneycalu I will take a look

File: src/ngMock/angular-mocks.js [hunk 2]
```diff
@@ -1125,7 +1125,76 @@ angular.mock.dump = function(object) {
...
+      .respond(function(method, url, data, headers, params) {
+        // for url of '/user/1234/article/567' params is {user: '1234', article: '567'}
+
+        angular.merge(MockArticleList[Number(params.article)], JSON.parse( data ));
+        MockUserList[Number(params.user)].lastUpdate = new Date();
+        return [201, MockArticleList[Number(params.article)]];
+      });
+  ```
+ *
+ * By default, query parameters on requested urls are parsed onto the params object, so a requsted url
```
Comment (gkalpak) [inline src/ngMock/angular-mocks.js]: _requsted_  -->  _requested_

File: src/ngMock/angular-mocks.js [hunk 3]
```diff
@@ -1125,7 +1125,76 @@ angular.mock.dump = function(object) {
...
+ * of `/list?q=searchstr&orderby=-name` would set params to be `{q: 'searchstr', orderby: '-name'}`
+ *
+ * For extra convenience, `whenRoute` and `expectRoute` shortcuts are available. These methods offer colon
+ * deliminated matching of solely the requrest url path, ignoring the query string. This allows declarations
```
Comment (gkalpak) [inline src/ngMock/angular-mocks.js]: _deliminated_  -->  _delimited_

File: test/ngMock/angular-mocksSpec.js [hunk 4]
```diff
@@ -963,6 +963,11 @@ describe('ngMock', function() {
       expect(typeof hb.whenHEAD).toBe("function");
     });
 
+    it('should provide "route" shortcuts for expect and when', function() {
+        expect(typeof hb.whenRoute).toBe("function");
+        expect(typeof hb.expectRoute).toBe("function");
```
Comment (gkalpak) [inline test/ngMock/angular-mocksSpec.js]: Too much indentation.

**Comment by petebacondarwin** [comment]
I did a bunch of tidy up on the docs and landed. Thanks @dmaloneycalu 
@gkalpak I missed that indentation. Sorry.

**Comment by dmaloneycalu** [comment]
Sorry on the indentation, was trying to be thorough checking since my editor is usually 4 spaces for indentation. Glad I could contribute.

**Comment by gkalpak** [comment]
:+1:

**Pull request closed without merging**